src/domain/tokens/entities/token.entity.ts

```diff
@@ -1,10 +1,45 @@
-import type { TokenSchema } from '@/domain/tokens/entities/schemas/token.schema';
-import type { z } from 'zod';
-
-export enum TokenType {
-  Erc721 = 'ERC721',
-  Erc20 = 'ERC20',
-  NativeToken = 'NATIVE_TOKEN',
-}
+import { z } from 'zod';
+import { buildPageSchema } from '@/domain/entities/schemas/page.schema.factory';
+import { AddressSchema } from '@/validation/entities/schemas/address.schema';
+
+const DEFAULT_ERC20_DECIMALS = 18;
+const DEFAULT_ERC721_DECIMALS = 0;
+
+const BaseTokenSchema = z.object({
+  address: AddressSchema,
+  logoUri: z.string().url(),
+  name: z.string(),
+  symbol: z.string(),
+  trusted: z.boolean(),
+});
+
+const NativeTokenSchema = BaseTokenSchema.extend({
+  type: z.literal('NATIVE_TOKEN'),
+  decimals: z.number(),
+});
+
+const Erc20TokenSchema = BaseTokenSchema.extend({
+  type: z.literal('ERC20'),
+  decimals: z.number().catch(DEFAULT_ERC20_DECIMALS),
+});
+
+const Erc721TokenSchema = BaseTokenSchema.extend({
+  type: z.literal('ERC721'),
```

Comment:
String `type`s are now preferred due to type complaints when using different enums from the domain vs. route layers.